File: src/main/java/core/procedures/system/Sleep.java
package core.procedures.system;

import core.exceptions.ThrowableWrapper;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Void;

public final class Sleep extends AFn {

  public Sleep() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[] {Long.class}).build());
  }

  @Override
  public String getName() {
    return "sleep";
  }

  @Override
  public Void apply1(Object arg) {
    try {
      Thread.sleep(((Number)arg).longValue());
    } catch (InterruptedException e) {
      throw new ThrowableWrapper(e);
    }
    return Void.VOID;
  }
}

File: src/main/java/core/procedures/system/Sleep.kt
package core.procedures.system

import core.procedures.AFn
import core.procedures.Arity.Range

class Sleep : AFn<Any?, Unit>(name = "sleep", arity = Range(0, 1), lastArgType = Number::class.java) {

    override fun invoke(args: Array<out Any?>) = when (args.size) {
        0    -> Thread.sleep(0)
        else -> Thread.sleep((args[0]!! as Number).toLong())
    }
}
